edward2/tensorflow/joint_distribution_test.py

```diff
@@ -0,0 +1,459 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for the JointDistributionEdward2."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import edward2 as ed
+
+import numpy as np
+
+import tensorflow.compat.v2 as tf
+import tensorflow_probability as tfp
+
+from tensorflow.python.framework import test_util  # pylint: disable=g-direct-tensorflow-import
+
+tfd = tfp.distributions
+
+
+@test_util.run_all_in_graph_and_eager_modes
+class JointDistributionEdward2Test(tf.test.TestCase):
+
+  def test_batch_and_event_shape_no_plate(self):
+    # The joint distribution specified below corresponds to this
+    # graphical model
+    #
+    #  (a)-->--(b)
+    #    \      |
+    #     \     v
+    #      `->-(c)
+
+    def dist():
+      a = ed.Bernoulli(probs=0.5, dtype=tf.float32)
+      b = ed.Bernoulli(probs=0.25 + 0.5*a, dtype=tf.float32)
+      ed.Normal(loc=a, scale=1. + b)
+
+    joint = ed.JointDistributionEdward2(dist, validate_args=True)
+
+    ds, _ = joint.sample_distributions()
+    self.assertLen(ds, 3)
+    self.assertIsInstance(ds[0], tfd.Bernoulli)
+    self.assertIsInstance(ds[1], tfd.Bernoulli)
+    self.assertIsInstance(ds[2], tfd.Normal)
+
+    is_event_scalar = joint.is_scalar_event()
+    self.assertAllEqual(is_event_scalar[0], True)
+    self.assertAllEqual(is_event_scalar[1], True)
+    self.assertAllEqual(is_event_scalar[2], True)
+
+    event_shape = joint.event_shape_tensor()
+    self.assertAllEqual(event_shape[0], [])
+    self.assertAllEqual(event_shape[1], [])
+    self.assertAllEqual(event_shape[2], [])
+
+    self.assertAllEqual(joint.event_shape, [[], [], []])
+
+    is_batch_scalar = joint.is_scalar_batch()
+    self.assertAllEqual(is_batch_scalar[0], True)
+    self.assertAllEqual(is_batch_scalar[1], True)
+    self.assertAllEqual(is_batch_scalar[2], True)
+
+    batch_shape = joint.batch_shape_tensor()
+    self.assertAllEqual(batch_shape[0], [])
+    self.assertAllEqual(batch_shape[1], [])
+    self.assertAllEqual(batch_shape[2], [])
+
+    self.assertAllEqual(joint.batch_shape, [[], [], []])
+
+  def test_batch_and_event_shape_with_plate(self):
+    # The joint distribution specified below corresponds to this
+    # graphical model
+    #
+    #       +-----------+
+    #  (g)--+-->--(loc) |
+    #       |       |   |
+    #       |       v   |
+    # (df)--+-->---(x)  |
+    #       +--------20-+
+
+    def dist():
```

Comment:
This is really interesting. Thanks Dave! I haven't following any discussions that led up to this invention. As a naive question, what's the advantage of writing a model like this? For example, does `dist` serve a purpose outside of being an argument to `JointDistributionEdward2`? More broadly, what are advantages of a JointDistribution as part of an Edward2 model, or as replacement of the current workflow of defining a generative process followed by `ed.make_log_joint`?